### PR TITLE
Add YouTube clone

### PR DESCRIPTION
**Pull request by SagarSDagdu**
### What's in this PR

Created YouTube clone target
Customized message list
Customization of individual messages according to YouTube live's UI

youtube_clone

### CODE REVIEW

File: YouTubeClone/SendButton/YTSendButton.swift [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        super.init(frame: frame)
+        // Customise the appearance to make it look like the YouTube SEND button
+        setImage(UIImage(systemName:"paperplane.fill"), for: .normal)
+        tintColor = .white
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+}
```
Comment (evsaev) [inline YouTubeClone/SendButton/YTSendButton.swift:24]:
```suggestion
final class YTSendButton: _Button {
    override func setUpAppearance() {
        // Customise the appearance to make it look like the YouTube SEND button
        setImage(UIImage(systemName:"paperplane.fill"), for: .normal)
        tintColor = .white
    }
}
```

File: YouTubeClone/YTLiveVideoViewController.swift [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+final class YTLiveVideoViewController: UIViewController {
+
+    private let url = URL(string: "https://devstreaming-cdn.apple.com/videos/streaming/examples/bipbop_adv_example_hevc/master.m3u8")!
+    
+    @IBOutlet private weak var videoView: UIView!
+    @IBOutlet private weak var chatView: UIView!
+    @IBOutlet private weak var playPauseButton: UIButton!
+    
+    @IBOutlet  var videoViewHeightConstraint: NSLayoutConstraint!
+    @IBOutlet  var videoViewFullScreenConstraint: NSLayoutConstraint!
```
Comment (evsaev) [inline YouTubeClone/YTLiveVideoViewController.swift:22]: Do we need an outlet for this constraint? Couldn't find the place where we change it

File: YouTubeClone/YTLiveVideoViewController.swift [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+final class YTLiveVideoViewController: UIViewController {
+
+    private let url = URL(string: "https://devstreaming-cdn.apple.com/videos/streaming/examples/bipbop_adv_example_hevc/master.m3u8")!
+    
+    @IBOutlet private weak var videoView: UIView!
+    @IBOutlet private weak var chatView: UIView!
+    @IBOutlet private weak var playPauseButton: UIButton!
+    
+    @IBOutlet  var videoViewHeightConstraint: NSLayoutConstraint!
```
Comment (evsaev) [inline YouTubeClone/YTLiveVideoViewController.swift:21]:
```suggestion
    @IBOutlet private var videoViewHeightConstraint: NSLayoutConstraint!
```

File: YouTubeClone/YTLiveVideoViewController.swift [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        videoPlayer.play()
+        isPlaying = true
+        playPauseButton.tintColor = .white
+        add(chatViewController, to: chatView)
+        chatViewController.view.translatesAutoresizingMaskIntoConstraints = false
+        chatView.addConstraints([
+            NSLayoutConstraint(item: chatViewController.view!, attribute: .top, relatedBy: .equal, toItem: chatView, attribute: .top, multiplier: 1, constant: 0),
+            NSLayoutConstraint(item: chatViewController.view!, attribute: .bottom, relatedBy: .equal, toItem: chatView, attribute: .bottom, multiplier: 1, constant: 0),
+            NSLayoutConstraint(item: chatViewController.view!, attribute: .leading, relatedBy: .equal, toItem: chatView, attribute: .leading, multiplier: 1, constant: 0),
+            NSLayoutConstraint(item: chatViewController.view!, attribute: .trailing, relatedBy: .equal, toItem: chatView, attribute: .trailing, multiplier: 1, constant: 0)
```
Comment (evsaev) [inline YouTubeClone/YTLiveVideoViewController.swift:52]: nit: layout anchors?

File: YouTubeClone/YTLiveVideoViewController.swift [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        ])
+        
+        setupUIAccordingToOrientation()
+    }
+    
+    override func viewDidLayoutSubviews() {
```
Comment (evsaev) [inline YouTubeClone/YTLiveVideoViewController.swift:58]: nit: super.viewDidLayoutSubviews()

File: YouTubeClone/YTLiveVideoViewController.swift [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        playerLayer.frame = videoView.bounds
+        playerLayer.videoGravity = .resizeAspect
+        playerLayer.removeFromSuperlayer()
+        videoView.layer.addSublayer(playerLayer)
```
Comment (evsaev) [inline YouTubeClone/YTLiveVideoViewController.swift:62]: can we add a player layer just once? 🤔

**Review by evsaev — APPROVED** [review]
Awesome 👏 Just a couple of nits

File: YouTubeClone/YTLiveVideoViewController.swift [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        videoView.bringSubviewToFront(playPauseButton)
+    }
+
+    override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
+        coordinator.animate { context in
+            self.setupUIAccordingToOrientation()
+        }
+    }
+    
+    @IBAction func playPauseButtonTapped(_ sender: Any) {
```
Comment (evsaev) [inline YouTubeClone/YTLiveVideoViewController.swift:72]: nit: private?

File: YouTubeClone/Composer/YTChatMessageComposerView.swift [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+import StreamChatUI
+import UIKit
+
+
+/// The customer composer view that tries to replicate the YouTube's composer view
+final class YTChatMessageComposerView: ComposerView {
+    lazy var dollarButton: UIButton = {
+        let button = UIButton()
+        button.setImage(UIImage(systemName: "dollarsign.square.fill"), for: .normal)
+        button.tintColor = .secondaryLabel
```
[the remark this answers is not in the document]
Comment (evsaev) [reply YouTubeClone/Composer/YTChatMessageComposerView.swift:19]: The tap handler is missing, right?

File: YouTubeClone/Composer/YTChatMessageComposerView.swift [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        NSLayoutConstraint.activate([
+            button.widthAnchor.constraint(equalToConstant: 30),
+            button.heightAnchor.constraint(equalToConstant: 30),
+        ])
```
[the remark this answers is not in the document]
Comment (SagarSDagdu) [reply YouTubeClone/Composer/YTChatMessageComposerView.swift]: Moved it to the right place 👍

File: YouTubeClone/Composer/YTChatMessageComposerView.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (SagarSDagdu) [reply YouTubeClone/Composer/YTChatMessageComposerView.swift]: Moved it to the right place, also added tap handlers in the VC 👍

File: YouTubeClone/Composer/YTChatMessageComposerView.swift [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        button.setImage(UIImage(systemName: "face.smiling.fill"), for: .normal)
+        button.tintColor = .secondaryLabel
+        NSLayoutConstraint.activate([
+            button.widthAnchor.constraint(equalToConstant: 30),
+            button.heightAnchor.constraint(equalToConstant: 30)
+            
+        ])
+        
+        return button
+    }()
```
[the remark this answers is not in the document]
Comment (SagarSDagdu) [reply YouTubeClone/Composer/YTChatMessageComposerView.swift]: Added the tap handlers

File: YouTubeClone/YTLiveVideoViewController.swift (hunk 4 above)
Comment (SagarSDagdu) [reply YouTubeClone/YTLiveVideoViewController.swift]: Changed this to use layout anchors

File: YouTubeClone/YTLiveVideoViewController.swift (hunk 6 above)
Comment (SagarSDagdu) [reply YouTubeClone/YTLiveVideoViewController.swift]: Done

File: YouTubeClone/YTLiveVideoViewController.swift (hunk 2 above)
Comment (SagarSDagdu) [reply YouTubeClone/YTLiveVideoViewController.swift]: Removed the unused constraint

File: YouTubeClone/YTChatMessageContentView.swift [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        
+        bubbleThreadMetaContainer.axis = .horizontal
+        
+        let subviews = bubbleThreadMetaContainer.subviews
+        bubbleThreadMetaContainer.removeAllArrangedSubviews()
+        bubbleThreadMetaContainer.addArrangedSubviews(subviews.reversed())
+        
+        let metadataSubviews = metadataContainer?.subviews
+        metadataContainer?.removeAllArrangedSubviews()
+        metadataContainer?.addArrangedSubviews((metadataSubviews?.reversed())!)
```
Comment (nuno-vieira) [inline YouTubeClone/YTChatMessageContentView.swift:52]: Probably worth to comment what are these lines doing 👍

**Review by nuno-vieira — CHANGES_REQUESTED** [review]
Overall Great Work 👏 Just one minor thing that needs to be changed, and then we can merge ✅

File: YouTubeClone/Composer/YTChatComposerViewController.swift [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+            .addTarget(self, action: #selector(showPayOptions(sender:)), for: .touchUpInside)
+    }
+        
+    override func showCommandSuggestions(for typingCommand: String) {
+        // Don't show suggestions
+    }
+
+    override func showMentionSuggestions(for typingMention: String, mentionRange: NSRange) {
+        // Don't show suggestions
+    }
```
[the remark this answers is not in the document]
Comment (SagarSDagdu) [reply YouTubeClone/Composer/YTChatComposerViewController.swift]: Thanks for this @nuno-vieira. Fixed this 👍🏻

File: YouTubeClone/YTChatMessageContentView.swift (hunk 11 above)
Comment (SagarSDagdu) [reply YouTubeClone/YTChatMessageContentView.swift]: I agree with you. Fixed this by adding descriptive comments regarding what those lines are doing.

**Review by nuno-vieira — APPROVED** [review]

**Pull request merged**